RevisionFilter: Indicate filters with button state and tooltip

gerhardol opened the pull request:
A follow up to #7598

## Proposed changes

#### RevisionFilter: Indicate filters with button state and tooltip

Visualize if there are filters available in the advanced filter.
Add tooltip with the current filter

#### RevDiff: Disable FilterInGrid if no selection
Exception occurred

#### RevisionFilter: Hotkey to disable filters

Ctrl-shift-F to disable the current filters (similar to what is done when switching repos)

Edi: Removed one commit that was moved to #9729 

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/6248932/141660697-88f3b5ac-a42f-416f-971c-6a136176a9f6.png)

![image](https://user-images.githubusercontent.com/6248932/141660930-76cc68c3-6cad-465e-af69-34ba357cd02e.png)

### After

![image](https://user-images.githubusercontent.com/6248932/141660772-efc27280-2d47-4c93-8a36-ec099caf508a.png)

![image](https://user-images.githubusercontent.com/6248932/141660780-dddf9d7b-29a6-4d81-a0ad-0790d65c0c43.png)

![image](https://user-images.githubusercontent.com/6248932/141660813-bf6ecbdb-b4d8-47fe-9c41-c6322267d352.png)


## Test methodology <!-- How did you ensure quality? -->

Manual

Adding TODO similar for GetRevisionFilter for FilterSummary
This is related to some other changes that should go into 4.0, so that should be done later.

## Merge strategy

- Rebase merge (PR submitter must change the commit message for the last commit).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitUI/UserControls/RevisionGrid/FilterInfo.cs:
```diff
@@ -339,6 +340,55 @@ public ArgumentString GetRevisionFilter()
             return filter;
         }
 
+        public string FilterSummary()
```
```suggestion
        public string GetSummary()
```

gerhardol replied:
Not changing in FilterChangedEventArgs.cs (unless requested)

RussKie replied:
In the filter it is a static string, hence the property. Here we run the method to build the summary.

RussKie reviewed (commented):
👍

gerhardol commented on UnitTests/GitUI.Tests/UserControls/FilterInfoTests.cs:
```diff
@@ -713,6 +713,6 @@ public void FilterInfo_DisableFilters_expected(bool byDateFrom, bool byDateTo, b
         }
 
-        // TODO: GetRevisionFilter
+        // TODO: GetRevisionFilter, FilterSummary
```
This is on TODO but I want this cleanup of these changes in 4.0
I can drop the tooltip too.

gerhardol commented on GitUI/UserControls/RevisionGrid/FormRevisionFilter.Designer.cs:
```diff
@@ -124,7 +124,6 @@ private void InitializeComponent()
             this.label1.Location = new System.Drawing.Point(3, 0);
             this.label1.Name = "label1";
             this.label1.Size = new System.Drawing.Size(67, 29);
-            this.label1.Text = "Since";
```
You do not see it I assume, but these strings do not have to be translated one extra time.

RussKie replied:
In the designer we need to have strings to visualise to a maintainer what's going on. If all labels are empty we're making it virtually impossible to maintain the form via the designer.
The strings in `InitializeComponent` are translated, if you may remember, so perhaps the summary can be built from the labels.

gerhardol replied:
How can the labels be exported?
Of course, the translation strings may be setup to only be used in the tooltip and the equality is tested...

gerhardol replied:
A solution is to add _NOTRANSLATION suffix to the labels and keep the string.

RussKie commented:
:+1:

On Tue, 16 Nov 2021 at 08:09, Gerhard Olsson ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitUI/UserControls/RevisionGrid/FormRevisionFilter.Designer.cs
> <https://github.com/gitextensions/gitextensions/pull/9728#discussion_r749683010>
> :
>
> > @@ -124,7 +124,6 @@ private void InitializeComponent()
>              this.label1.Location = new System.Drawing.Point(3, 0);
>              this.label1.Name = "label1";
>              this.label1.Size = new System.Drawing.Size(67, 29);
> -            this.label1.Text = "Since";
>
> A solution is to add _NOTRANSLATION suffix to the labels and keep the
> string.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9728#discussion_r749683010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXTBJQRNSZ62HE6PMHDUMFZIPANCNFSM5H7EQ25A>
> .
>

gerhardol commented:
Squashed, added a fixup to refresh after disabling the filters.
Plan to merge tomorrow, maybe with a simple test (including GetRevisionFilter), otherwise that is for TODO.

gerhardol commented:
Added a very simple test and added a fix for off by one in #9733
Squashing and merging tomorrow
